File: App/app/src/main/java/com/runflow/app/data/model/AuthModels.kt
package com.runflow.app.data.model

import kotlinx.serialization.Serializable

@Serializable
data class LoginRequest(
    val code: String
)

@Serializable
data class LoginResponse(
    val accessToken: String,
    val refreshToken: String,
    val user: User
)

@Serializable
data class RefreshTokenRequest(
    val refreshToken: String
)

@Serializable
data class RefreshTokenResponse(
    val accessToken: String,
    val refreshToken: String
)

@Serializable
data class LogoutRequest(
    val refreshToken: String
)

@Serializable
data class AuthResponse(
    val success: Boolean
)

@Serializable
data class User(
    val id: String,
    val name: String? = null,
    val email: String? = null,
    val image: String? = null,
    val sex: Sex? = null,
    val birthDate: String? = null,
    val hrMax: Int? = null,
    val hrRest: Int? = null,
    val weight: Float? = null,
    val height: Float? = null,
    val hrZone1Max: Int? = 130,
    val hrZone2Max: Int? = 140,
    val hrZone3Max: Int? = 150,
    val hrZone4Max: Int? = 160,
    val vdotCorrectionFactor: Float? = 1.0f,
    val lastSyncAt: String? = null,
    val syncInProgress: Boolean = false
)

@Serializable
enum class Sex {
    MALE,
    FEMALE,
    OTHER
}

@Serializable
data class EmailLoginRequest(
    val email: String,
    val password: String
)

@Serializable
data class RegisterRequest(
    val email: String,
    val password: String,
    val name: String? = null
)

File: flutter/lib/data/models/auth_models.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:runflow_flutter/data/models/json_compat.dart';

part 'auth_models.freezed.dart';
part 'auth_models.g.dart';

@Freezed(copyWith: true)
sealed class LoginRequest with _$LoginRequest {
  const factory LoginRequest({
    required String code,
    String? redirectUri,
  }) = _LoginRequest;
  const LoginRequest._();

  factory LoginRequest.fromJson(Map<String, dynamic> json) =>
      _$LoginRequestFromJson(json);
}

@Freezed(copyWith: true)
sealed class LoginResponse with _$LoginResponse {
  const factory LoginResponse({
    required String accessToken,
    required String refreshToken,
    required int expiresIn,
    required String tokenType,
    required User user,
  }) = _LoginResponse;
  const LoginResponse._();

  factory LoginResponse.fromJson(Map<String, dynamic> json) =>
      _$LoginResponseFromJson(json);
}

@Freezed(copyWith: true)
sealed class RefreshRequest with _$RefreshRequest {
  const factory RefreshRequest({required String refreshToken}) = _RefreshRequest;
  const RefreshRequest._();

  factory RefreshRequest.fromJson(Map<String, dynamic> json) =>
      _$RefreshRequestFromJson(json);
}

@Freezed(copyWith: true)
sealed class RefreshResponse with _$RefreshResponse {
  const factory RefreshResponse({
    required String accessToken,
    required String refreshToken,
  }) = _RefreshResponse;
  const RefreshResponse._();

  factory RefreshResponse.fromJson(Map<String, dynamic> json) =>
      _$RefreshResponseFromJson(json);
}

enum Sex { male, female, other }

@Freezed(copyWith: true)
sealed class User with _$User {
  const factory User({
    required String id,
    String? email,
    String? name,
    String? image,
    @JsonKey(fromJson: sexFromJson, toJson: sexToJson) Sex? sex,
    @JsonKey(fromJson: flexibleDateTimeFromJson, toJson: dateTimeToJson)
    DateTime? birthDate,
    int? hrMax,
    int? hrRest,
    double? weight,
    double? height,
    double? vdotCorrectionFactor,
    @JsonKey(fromJson: flexibleDateTimeFromJson, toJson: dateTimeToJson)
    DateTime? lastSyncAt,
  }) = _User;
  const User._();

  factory User.fromJson(Map<String, dynamic> json) => _$UserFromJson(json);
}

@Freezed(copyWith: true)
sealed class ApiError with _$ApiError {
  const factory ApiError({
    required String error,
    required DateTime timestamp,
    String? code,
    Map<String, dynamic>? details,
    String? path,
  }) = _ApiError;
  const ApiError._();

  factory ApiError.fromJson(Map<String, dynamic> json) =>
      _$ApiErrorFromJson(json);
}

Sex? sexFromJson(Object? value) {
  return switch (compatibilitySexFromJson(value)) {
    CompatibilitySex.male => Sex.male,
    CompatibilitySex.female => Sex.female,
    CompatibilitySex.other => Sex.other,
    null => null,
  };
}

String? sexToJson(Sex? value) {
  return switch (value) {
    Sex.male => compatibilitySexToJson(CompatibilitySex.male),
    Sex.female => compatibilitySexToJson(CompatibilitySex.female),
    Sex.other => compatibilitySexToJson(CompatibilitySex.other),
    null => null,
  };
}
